tests/run-make/no-alloc-shim/rmake.rs

```diff
@@ -0,0 +1,36 @@
+// This test checks the compatibility of the interaction between `--emit obj` and
+// `#[global_allocator]`, as it is now possible to invoke the latter without the
+// allocator shim since #86844. As this feature is unstable, it should fail if
+// --cfg check_feature_gate is passed.
+// See https://github.com/rust-lang/rust/pull/86844
+
+//@ ignore-cross-compile
+// Reason: the compiled binary is executed
+
+//FIXME(Oneirical): ignore-msvc FIXME(bjorn3) can't figure out how to link with the MSVC toolchain
```

Comment:
Discussion: I'm fine with ignoring msvc for now and leaving this FIXME in, unless you want to investigate how to make msvc linking succeeds here.

Reply:
I suspect that the C runtime library isn't being passed to the linker because the compiler isn't being given any source code to compile, only object files. Adding `libcmt.lib` to the args (if msvc) would fix this but maybe there should be some more principled way of doing this. Maybe a `link()` alternative to `cc` that only does linking? Not sure.

One odd thing I noticed is `__udivti3`. That suggests to me that compiler-builtins is also needed.

Reply:
Oh, actually we already have [`extra_c_flags`](https://github.com/rust-lang/rust/blob/176e5452095444815207be02c16de0b1487a1b53/src/tools/run-make-support/src/external_deps/cc.rs#L123C8-L123C21). Instead of using a hard coded list, maybe we could just use [`--print native-static-libs`](https://doc.rust-lang.org/rustc/command-line-arguments.html#--print-print-compiler-information) to grab all the necessary ones. That should also include the CRT. But that's probably a big enough job to do separately.

Reply:
I was able to get this test to pass on msvc by adding `libcmt.lib` as Chris mentioned, but also had to pass compiler-builtins:

```rs
use run_make_support::{
    cc, cwd, extra_c_flags, has_extension, has_prefix, run, rustc, shallow_find_files,
};

fn main() {
    rustc().input("foo.rs").crate_type("bin").emit("obj").panic("abort").run();
    let libdir = rustc().print("target-libdir").run().stdout_utf8();
    let libdir = libdir.trim();
    let alloc_libs = shallow_find_files(&libdir, |path| {
        has_prefix(path, "liballoc-") && has_extension(path, "rlib")
    });
    let core_libs = shallow_find_files(libdir, |path| {
        has_prefix(path, "libcore-") && has_extension(path, "rlib")
    });
    let compiler_builtins_libs = shallow_find_files(libdir, |path| {
        has_prefix(path, "libcompiler_builtins") && has_extension(path, "rlib")
    });

    cc().args(extra_c_flags())
        .input("foo.o")
        .out_exe("foo")
        .arg("/link")
        .args(&alloc_libs)
        .args(&core_libs)
        .args(&compiler_builtins_libs)
        .arg("libcmt.lib")
        .run();
    run("foo");

    // Check that linking without __rust_no_alloc_shim_is_unstable defined fails
    rustc()
        .input("foo.rs")
        .crate_type("bin")
        .emit("obj")
        .panic("abort")
        .cfg("check_feature_gate")
        .run();
    cc().args(extra_c_flags())
        .input("foo.o")
        .out_exe("foo")
        .arg("/link")
        .args(&alloc_libs)
        .args(&core_libs)
        .args(&compiler_builtins_libs)
        .arg("libcmt.lib")
        .run_fail();
}
```

Not super sure about what's the best thing to do w.r.t. having to pass `libcmt.lib`.

Reply:
Maybe it's better to accept this PR with `ignore-msvc` and then later figure the thing to do?

Reply:
> Instead of using a hard coded list, maybe we could just use [`--print native-static-libs`](https://doc.rust-lang.org/rustc/command-line-arguments.html#--print-print-compiler-information) to grab all the necessary ones. That should also include the CRT. But that's probably a big enough job to do separately.

I briefly looked at this and I don't think this is possible (currently), because its docs say:

> This may be used when creating a staticlib crate type. If this is the only flag, it will perform a full compilation and include a diagnostic note that indicates the linker flags to use when linking the resulting static library.

Unfortunately, we're building a bin here, and I tried this locally, `--print native-static-libs` did not emit anything (just a newline).

Reply:
> Maybe it's better to accept this PR with `ignore-msvc` and then later figure the thing to do?

Yeah, this is was what I had in mind in https://github.com/rust-lang/rust/pull/128407#discussion_r1702439172, and I don't want to block this PR on trying to unignore it on msvc -- I can try to investigate that separately.

Reply:
@Oneirical let's add a `ignore-msvc` for this with a FIXME linking to #128602 and call it a day for now 😆 I'll investigate separately.

@rustbot author

Reply:
> Unfortunately, we're building a bin here, and I tried this locally, --print native-static-libs did not emit anything (just a newline).

You'd need to build a fake staticlib then parse stdout. I think this could be done once in the `extra_c_flags` function and cached instead of hard coding a list of libraries to link, Or maybe it could be something compiletest does.

Reply:
Very well. I have restored the ignore:

```
//@ ignore-msvc
//FIXME(Oneirical): Getting this to work on MSVC requires passing libcmt.lib to CC,
// which is not trivial to do.
// Tracking issue: https://github.com/rust-lang/rust/issues/128602
// Discussion: https://github.com/rust-lang/rust/pull/128407#discussion_r1702439172
```

@rustbot review